c7n/offhours.py

```diff
@@ -277,32 +293,164 @@ def get_local_tz(self, parts):
         return tz
-    
+
+
+    def is_custom(self, parts):
+        return "tz" in parts and "on" in parts and "off" in parts
+
+
+    def get_custom_time(self, i, now, parts):
+        if self.time_type is None:
+            return False
+
+        time = parts.get(self.time_type)
+        tz = parts.get("tz")
+
+        for item in time:
+            days = item.get("days")
+            hour = item.get("hour")
+            for d in days:
+                log.debug("resource: %s checking custom hours, day=%s "
+                     "hour=%s, and tz=%s" % (resource_id(i), d, hour, tz))
+                if now.weekday() == VALID_DAYS.index(d): #if today
+                    if now.hour == hour:
+                        log.debug("resource: %s matches custom hours with "
+                                  "value now: %s" % (resource_id(i), now))
+                        return True
+        log.debug("No custom hours match found for resource: %s" % resource_id(i))
+        return False
+
 
 class OffHour(Time):
 
     schema = type_schema(
         'offhour', rinherit=Time.schema, required=['offhour', 'default_tz'],
         offhour={'type': 'integer', 'minimum': 0, 'maximum': 24})
+    time_type = "off"
 
     def get_sentinel_time(self, tz):
         t = super(OffHour, self).get_sentinel_time(tz)
         return t.replace(hour=self.data.get('offhour', DEFAULT_OFFHOUR))
 
-                         
+
 class OnHour(Time):
 
     schema = type_schema(
         'onhour', rinherit=Time.schema, required=['onhour', 'default_tz'],
         onhour={'type': 'integer', 'minimum': 0, 'maximum': 24})
+    time_type = "on"
 
     def get_sentinel_time(self, tz):
         t = super(OnHour, self).get_sentinel_time(tz)
         return t.replace(hour=self.data.get('onhour', DEFAULT_ONHOUR))
 
-    
-    
 
+
+class ScheduleParser:
+
+    cache = {}
+
+    def parse(self, tag_value):
+        # check the cache
+        if tag_value in self.cache:
+            return self.cache[tag_value]
+        schedule = {}
+        # parse schedule components
+        pieces = tag_value.split(';')
+        for piece in pieces:
+            kv = piece.split('=')
+            # components must by key-value
+            if not len(kv) == 2:
+                continue
+            key = kv[0].strip()
+            value = kv[1].strip()
+            if key == 'on' or key == 'off':
+                # parse custom on/off hours
+                value = self.parse_custom_hours(value)
+            schedule[key] = value
+        # add default timezone, if none supplied or blank
+        if schedule.get('tz') is None  or not schedule.get('tz'):
+            schedule['tz'] = DEFAULT_TZ
+        # validate
+        if not self.is_valid(schedule):
+            schedule = None
+        # cache
+        self.cache[tag_value] = schedule
+        return schedule
+
+    def parse_custom_hours(self, hours):
+        parsed = []
+        hours = hours.translate(None, '[]').split(',(')
```

Comment:
interesting, hadn't seen a None param to translate before.

The split ',(' seems a little fragile